Build a wallet using js - JS Bounty (#1752)
XRPLF/xrpl-dev-portal: content/tutorials/build-apps/build-a-desktop-wallet-in-js.md

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"
SEED="sEd7B2Jm7RBhhh5i88SH38aZB8ot8Kq" // Replace with your seed
```
6. Change the seed to your own seed. You can get a testnet account from [here](https://xrpl.org/xrp-test-net-faucet.html).

### Step 2: Running the App

In this step, we will set up the vite bundler to run the application. We've used Vite to bundle the app, if you're unfamiliar with Vite, you can check out the [Vite documentation](https://vitejs.dev/guide/) to learn more.

1. To set up vite bundler, copy the code written below to the vite.config.js file. xrpl.js requires some configuration to run with the vite bundler. You can find the configuration in the code below or you can check out the [xrpl.js documentation](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).

{{ include_code("_code-samples/build-a-wallet/js/vite.config.js", language="js") }}

2. Add script to package.json

In your package.json file, add:

```json
"scripts": {
    "dev": "vite"
}

> ```

Reviewer: I'd recommend removing this section divider, and just including these as additional steps in the 1st step of "Setting up the project"

Now the basic setup of your application is ready, after completing the next steps of this tutorial you can run this command

```bash
yarn dev ## or npm run dev
```

in your terminal which will start your app in development mode. Your terminal will display an URL using which you can verify your changes in real-time.

> If you get stuck building while doing this tutorial, or working on another project, feel free to ask for help in the XRPL's [Developer Discord](https://discord.com/invite/KTNmhJDXqa).

Reviewer: If this works right away, I'd keep this here, otherwise I'd move this comment about using yarn dev or npm run dev until you know it'll work. (e.g. after completing the 1st set of steps)

I'd also separate the `yarn dev` and `npm run dev` into two separate lines in order to make it harder to miss the npm run dev instruction, e.g.
`yarn dev`
or
`npm run dev`

Author: I think it'll be better to place this as a last step in `step 2` after the home page code is done.

### Step 3: Creating the Home Page (Displaying Account & Ledger Details)

In this step, we will create the home page that will display the account details and the ledger details.

![Home Page Screenshot](img/js-wallet-home.png)

> 1. Make a new folder named src in the root directory of the project.

Reviewer: ```suggestion
1. Make a new folder named `src` in the root directory of the project.
```

> 2. Create new files named index.html, index.js and index.css. 

Reviewer: Please apply this to all filenames (you can search for `.html`, `.js` and `.css`)
```suggestion
2. Create new files named `index.html`, `index.js` and `index.css`. 
```

Author: Added :)

3. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html), [index.js](_code-samples/build-a-wallet/js/index.js) and [index.css](_code-samples/build-a-wallet/js/index.css) files to your files.

> Now that you've got the basic setup, let's discuss what we're going to do. For the home page, we want to display our account info, what's happening on the ledger, and for fun add a little logo. In order to make that happen, first we need to look up our account state. These next few steps will walk you through that process.

Reviewer: Shifting to a bulleted list rather than comma separated list because it's easier to read.

```suggestion
Now that you've got the basic setup, let's discuss what we're going to do. For the home page, we want to:

- Display our account info
- Show what's happening on the ledger
- And add a little logo for fun

In order to make that happen, we need to look up our account state on the ledger. The next couple steps will walk you through that process.
```

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/submit-transaction.js", language="js") }}

1. Now back to the send-xrp.js file, copy the code written below to the file. In this piece of code we are first getting all the DOM elements from HTML and adding event listners to update & validate the fields based on the user input. Using `renderAvailableBalance` method we display the current available balance of the wallet. `validateAddress` function validates the user address, and the amount is validated using a regular expression. When all the fields are filled with correct inputs, we call the `submitTransaction` function to submit the transaction to the ledger.

{{ include_code("_code-samples/build-a-wallet/js/src/send-xrp/send-xrp.js", language="js") }}

### Step 5: Creating the Transactions Page

> Now that we have created the home page and the send XRP page, let's create the transactions page that will display the transaction history of the account.

Reviewer: GIving context to the big list
```suggestion
Now that we have created the home page and the send XRP page, let's create the transactions page that will display the transaction history of the account. In order to see what's happening on the ledger, we're going to display the following fields: 
```

  - Account: The account that sent the transaction.
  - Destination: The account that received the transaction.
  - Amount: The amount of XRP sent in the transaction.
  - Transaction Type: The type of transaction.
  - Result: The result of the transaction.
  - Link: A link to the transaction on the XRP Ledger Explorer.

![Transactions Page Screenshot](img/js-wallet-transaction.png)

1. Create a folder named `transaction-history` in the src directory.

> 
> 2. Add two files named `transaction-history.js` and `transaction-history.html`.
> 
> 3. In the `transaction-history.js` file, copy the code written below. To fetch the transactions from XRP Ledger we are using [account_tx](https://xrpl.org/account_tx.html) method. This method returns a list of transactions that affected the specified account. We are using the `limit` parameter to limit the number of transactions to 10 per page. We are also using the `marker` parameter to paginate through the transactions.
> 
> {{ include_code("_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.js", language="js") }}
> 
> 1. Copy the contents of the [transaction-history.html](_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.html) file to your transaction-history.html file. In this file we are defining a table  which displays the mentioned fields mentioned above.

Reviewer: Two things here:
1. Creating the file should happen in the same step as "copying this code"
2. The additional text giving context should be on a separate line from the step (Generally it's better to have the steps be focused on actions, with any surrounding text giving more information once they've completed the task)
```suggestion
2. Create a file named `transaction-history.js` and copy the code written below. 

{{ include_code("_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.js", language="js") }}

This code uses [account_tx](https://xrpl.org/account_tx.html) to fetch transactions we've sent to and from this account. In order to get all the results, we're using the `marker` parameter to paginate through the incomplete list of transactions until we reach the end.

3. Create a file named "transaction-history.html" and copy the code from [transaction-history.html](_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.html) into it. 

`transaction-history.html` defines a table which displays the fields mentioned above.
```

In the function `getWalletDetails`, we are using the [account_info](https://xrpl.org/account_info.html) method to fetch the account details. We are also using the [server_info](https://xrpl.org/server_info.html) method to calculate the reserve base and increment. You can read more about the reserve requirement [here](https://xrpl.org/reserves.html).

5. Create another file in helpers folder named get-ledger-details.js which will be used to fetch the ledger details and copy the code written below to the file.

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-ledger-details.js", language="js") }}

In the function `getLedgerDetails`, we are using the [ledger](https://xrpl.org/ledger.html) that will fetch the details of the latest ledger, you can tweak the code to fetch the details of a specific ledger and also to fetch account details and other information.

6. Now, let's add the code to index.js file to fetch the account and ledger details and display them on the home page. Copy the code written below to the index.js file.

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

> In the code above, we imported the `getWalletDetails` and `getLedgerDetails` functions from the get-wallet-details.js and get-ledger-details.js files respectively. Added event listners to the buttons for redirection. Used the `setInterval` method to call the getLedgerDetails function every 4 seconds. (Note: The ledger's estimate close time is 4 seconds, so we are calling the function every 4 seconds to get the latest ledger details. This can be adjusted according to your needs.)

Reviewer: It's better to describe the what/why rather than the implementation here. So focusing on what we're trying to accomplish, and if one of the helpers is hard to understand giving a bit more context. Ex of how to do that suggested here.

```suggestion
In the code above we rendered the wallet and ledger details using the information we got earlier in `get-wallet-details.js` and `get-ledger-details.js`. In order to make sure we have up to date data, we call the `getLedgerDetails` function every 4 seconds, as that's about how long it takes for a ledger to close on mainnet. 

The one other thing we do here is add two buttons - one to send XRP and one to view the transaction history of our account. They won't work just yet - we'll have to add those implementations next.
```

Up next we need to update the HTML and CSS for the home page to give it the XRPL-look and feel. 

7. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html) file to your index.html file.

For this tutorial, we're using basic HTML and pre-loaded CSS files. If you'd like to change the aesthetic, you can replace it with any CSS framework of your choice!

> 8. Copy the code from the [index.css](_code-samples/build-a-wallet/js/index.css) file to your project's index.css file to use the pre-made stylings.

Reviewer: I think we already did this in step 3

### Step 3: Creating the Send XRP Page

> Till now, we have created the home page that displays the account and ledger details. In this step, we will create the Send XRP page that will allow users to send XRP to another account.

Reviewer: Focusing on what they will accomplish here, rather than what they'll implement.
```suggestion
Now that we've created the home page, we can move on to the "Send XRP" page. This is what allows this wallet to manage your account's funds.
```

![Send XRP Page Screenshot](img/js-wallet-send-xrp.png)

> 1. Create a folder named send-xrp in the src directory.

Reviewer: nit: Here and elsewhere, please add the backticks when you're referencing a file, function, or code. 

```suggestion
1. Create a folder named `send-xrp` in the `src` directory.
```

> 2. Add two files named send-xrp.js and send-xrp.html. You can also create a separate css file depending on your choice. 

Reviewer: It's a bit unclear where you wanted these files created.

Also, I think it's alright not to mention the css changes, as that's not really the point of the tutorial. (You've already let them know that they can challenge themselves to add their own css - if they're taking on that challenge they probably don't need much more guidance on how to set it up)

```suggestion
2. Inside the `send-xrp` folder, create two files named send-xrp.js and send-xrp.html.
```

3. Copy the contents of the [send-xrp.html](_code-samples/build-a-wallet/js/src/send-xrp/send-xrp.html) file to your send-xrp.html file. The provided HTML code includes three input fields for the destination address, amount, and destination tag, each with their corresponding labels.

> 4. Now that we have the HTML code, let's add the JavaScript code. In helpers folder, add a new file named submit-transaction.js and copy the code written below to the file.

Reviewer: Similar to the previous suggestion, it's helpful to describe what the code you're copying is going to accomplish before you ask them to copy it. 

Also, changing "add" to "create" if we're making a new file because "add" could mean to copy an existing file. 

```suggestion
4. Now that we have the HTML code, let's add the JavaScript code. In the `helpers` folder, create a new file named `submit-transaction.js` and copy the code written below to the file.
```

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/submit-transaction.js", language="js") }}

> In the code above, we are using the [submit](https://xrpl.org/submit.html) method to submit the signed transaction to the XRPL. You can read more about the submit method [here](https://xrpl.org/submit.html).

Reviewer: Since you link to submit already, you don't need a second sentence explaining where to find `submit` - it's implied that if they want to read more they can follow the link :) 

(Please apply this to the rest of the file)

```suggestion
In the code above, we are using the [submit](https://xrpl.org/submit.html) method to submit the signed transaction to the XRPL.
```

Reviewer: I think it'd also be worth elaborating a bit here about how we submit this to the ledger. (E.g. How are we able to sign it?)

Author: Added

5. Now back to the send-xrp.js file, copy the code written below to the file.

{{ include_code("_code-samples/build-a-wallet/js/src/send-xrp/send-xrp.js", language="js") }}

> In the code above, we retrieve elements from DOM, add event listeners, validate inputs and after validation, we call the `submitTransaction` function from the submit-transaction.js file to submit the transaction to the XRPL.

Reviewer: This describes what the code is doing, but doesn't explain why. What functional pieces of the wallet should the user expect to work after they copy this in?

Author: Added few more things in this description

### Step 4: Creating the Transactions Page

Now that we have created the home page and the send XRP page, let's create the transactions page that will display the transaction history of the account.

>   1. Account: The account that sent the transaction.

Reviewer: The screenshot is a bit confusing because the table doesn't display the "Destination" - Could you update the table so:
1. Instead of `Account` it shows `Destination`
2. Somewhere on the screen the user can see their `Account`? 

That way all the key information is on the page. (This may involve modifying the above code snippets to match the implementation)

Author: Thanks for pointing it out, forgot to update the screenshot. Fixed.

  2. Destination: The account that received the transaction.
  3. Amount: The amount of XRP sent in the transaction.
  4. Transaction Type: The type of transaction.
  5. Result: The result of the transaction.

>   6. Link: A link to the transaction on the XRP Ledger Explorer.

Reviewer: Maybe instead of numbers, this section should be bullets? That way it's not as confusing when you scroll down and see the numbers restart at 1. because there are instructions to follow.

Author: Added

![Transactions Page Screenshot](img/js-wallet-transaction.png)

1. Create a folder named transaction-history in the src directory.

2. Add two files named transaction-history.js and transaction-history.html.

3. In the transaction-history.js file, copy the code written below.

> {{ include_code("_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.js", language="js") }}

Reviewer: Ditto on explaining the intent before copying the code.

Author: Added

In the code above we fetch the transactions from the XRP Ledger by using the [account_tx](https://xrpl.org/account_tx.html) method. You can read more about the account_tx method [here](https://xrpl.org/account_tx.html).

> 4. Copy the contents of the [transaction-history.html](_code-samples/build-a-wallet/js/src/transaction-history/transaction-history.html) file to your transaction-history.html file.

Reviewer: What does this accomplish?

Author: Added another sentence "In this file we are defining a table  which displays the mentioned fields mentioned above."

> You can use this code as a starting point and customize it to suit your needs. For example, you might want to add more columns to the table. You can also check out the [XRP Ledger Explorer](https://livenet.xrpl.org/) to see how the transaction details are displayed.

Reviewer: Changing to be a specific challenge.
```suggestion
You can use this code as a starting point for displaying your account's transaction history. If you want an additional challenge, try expanding it to support different transaction types (e.g. [TrustSet](https://xrpl.org/trustset.html)). If you want inspiration for how to handle this, you can check out the [XRP Ledger Explorer](https://livenet.xrpl.org/) to see how the transaction details are displayed.
```

> ### Step 5: Running the App

Reviewer: I think we should do this earlier - it's better if the user is able to see the results of each step after they do it, and so having them do the necessary setup near the beginning let's them feel proud of each step, rather than waiting until the very end to play with what they're building.

Author: It's Step 2 now

A wallet app is not complete without a way to run it. In this step, we will set up the vite bundler to run the app.

1. Set up vite bundler, copy the code written below to the vite.config.js file.

{{ include_code("_code-samples/build-a-wallet/js/vite.config.js", language="js") }}

We've used Vite to bundle the app, if you're unfamiliar with Vite, you can check out the [Vite documentation](https://vitejs.dev/guide/) to learn more.

xrpl.js requires some configuration to run with the vite bundler. You can find the configuration in the code below or you can check out the [xrpl.js documentation](https://github.com/XRPLF/xrpl.js/blob/main/UNIQUE_SETUPS.md#using-xrpljs-with-vite-react).

2. Add script to package.json

In your package.json file, add:

```json
"scripts": {
    "dev": "vite"
}
```

Now our app is ready to run. To run the app, open terminal in the project directory and run:

```bash
yarn dev ## or npm run dev
```

This will start the app in development mode. Check your terminal for the URL where the app is running. Click the URL to open the app in your browser. You should see the home page of the app. 

> If you ran into trouble while running the app, go over the steps again to ensure you followed them correctly. If still the app doesn't run, feel free to ask on XRPL Community [Discord](https://discord.com/invite/KTNmhJDXqa).

Reviewer: I'd recommend removing the suggestion to re-read the steps because that can seem a bit condescending (it's usually what people try first)

```suggestion
If you get stuck building while doing this tutorial, or working on another project, feel free to ask for help in the XRPL's [Developer Discord](https://discord.com/invite/KTNmhJDXqa).
```

## Next Steps

Now that you have a functional wallet, you can take it in several new directions. The following are a few ideas:

- You could support more of the XRP Ledger's [transaction types](transaction-types.html) including [tokens](issued-currencies.html) and [cross-currency payments](cross-currency-payments.html)

> - You could add support for multiple currencies, including IOUs.

Reviewer: IOU's are an old name for Tokens :)
```suggestion
- You could add support for displaying multiple tokens, beyond just XRP
```

> - Allow the user to trade in the [decentralized exchange](decentralized-exchange.html)

Reviewer: ```suggestion
- You could support creating [offers](https://xrpl.org/offers.html) in the [decentralized exchange](decentralized-exchange.html)
```

> - Add a way to request payments, such as with QR codes or URIs that open in your wallet.

Reviewer: ```suggestion
- You could add new ways to request payments, such as with QR codes or URIs that open in your wallet.
```

> - Support better account security including [regular key pairs](cryptographic-keys.html#regular-key-pair) or [multi-signing](multi-signing.html).

Reviewer: ```suggestion
- Or you could support better account security including allowing users to set [regular key pairs](cryptographic-keys.html#regular-key-pair) or handle [multi-signing](multi-signing.html).
```

  
4. Run `yarn` or `npm i` to install the packages.

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"
SEED="sEd7B2Jm7RBhhh5i88SH38aZB8ot8Kq" // Replace with your seed
```
6. Change the seed to your own seed. You can get a testnet account from [here](https://xrpl.org/xrp-test-net-faucet.html).

### Step 2: Creating the Home Page (Displaying Account & Ledger Details)

In this step, we will create the home page that will display the account details and the ledger details.

![Home Page Screenshot](img/js-wallet-home.png)

1. Make a new folder named src in the root directory of the project.

2. Create new files named index.html, index.js and index.css. 

3. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html), [index.js](_code-samples/build-a-wallet/js/index.js) and [index.css](_code-samples/build-a-wallet/js/index.css) files to your files.

> 4. Create a new file in helpers folder named get-wallet-details.js and copy the code written below to the file.

Reviewer: This is a lot of code to copy and paste - since this snippet is one that you're hoping the reader will be able to understand, I'd recommend taking a bit more time before showing the code to explain what we're about to do and why. 

Ex. 

"Now that you've got the basic setup, let's discuss what we're going to do. For the home page, we want to display our account info, what's happening on the ledger, and for fun add a little logo. In order to make that happen, first we need to look up our account state. This next bit of code handles that.

4. Create a new file in your `helpers` folder named `get-wallet-details.js` and copy the code written below into the file. "

...

(Later on before step 5)

"Now that we've got our account information, we want to look up the latest ledger to verify that the data we're showing people is up to date."

5. Create another file...

Reviewer: You can totally explain the exact technical details of how it works after, but by giving people context on what's about to happen + why it matters they'll have a better understanding of the important pieces of the overall design :)

Author: Got it, changed the wordings

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-wallet-details.js", language="js") }}

> In the function getWalletDetails, we are using the [account_info](https://xrpl.org/account_info.html) method to fetch the account details. We are also using the [server_info](https://xrpl.org/server_info.html) method to calculate the reserve base and increment. You can read more about the reserve requirement [here](https://xrpl.org/reserves.html). Follow the comments in the code to understand the logic.

Reviewer: nit: Adding the code backticks for code names.
```suggestion
In the function `getWalletDetails`, we are using the [account_info](https://xrpl.org/account_info.html) method to fetch the account details. We are also using the [server_info](https://xrpl.org/server_info.html) method to calculate the reserve base and increment. You can read more about the reserve requirement [here](https://xrpl.org/reserves.html). Follow the comments in the code to understand the logic.
```

> In the function getWalletDetails, we are using the [account_info](https://xrpl.org/account_info.html) method to fetch the account details. We are also using the [server_info](https://xrpl.org/server_info.html) method to calculate the reserve base and increment. You can read more about the reserve requirement [here](https://xrpl.org/reserves.html). Follow the comments in the code to understand the logic.

Reviewer: Also, instead of saying "follow the comments" after the code, you'll probably want to say that beforehand. Otherwise the reader would have to scroll down, then back up, as opposed to just paying attention as they scroll down :)

Author: Removed, English is the hardest part of coding :p

{{ include_code("_code-samples/build-a-wallet/js/src/helpers/get-ledger-details.js", language="js") }}

In the function getLedgerDetails, we are using the [ledger](https://xrpl.org/ledger.html) that will fetch the details of the latest ledger, you can tweak the code to fetch the details of a specific ledger and also to fetch account details and other information. Follow the link to read more about the ledger method.

6. In the index.js file, import the getWalletDetails and getLedgerDetails functions from the helpers folder.

Your index.js file should look like this:

{{ include_code("_code-samples/build-a-wallet/js/index.js", language="js") }}

In index.js file, define the variables that get the elements from the DOM and then add event listeners to the buttons using the javascript addEventListener method you read more about this method [here](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener). On clicking the buttons you'll be redirected to the Send XRP page and Transactions page respectively which we'll create in the next steps of this tutorial. 

> Call the getWalletDetails and getLedgerDetails functions to fetch the account and ledger details respectively. We are using the setInterval method to call the getLedgerDetails function every 4 seconds. You can read learn about the setInterval method from this [link](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval). (Note: The ledger's estimate close time is 4 seconds, so we are calling the function every 4 seconds to get the latest ledger details. This can be adjusted according to your needs.)

Reviewer: I'm not sure if this is an instruction, or if it's just describing something. What is the action you expect the user to take here if any?

Author: I wanted to add a bit of description, but now have removed a lot of jargon and have kept everything simple.

**Optional:**

In the code above you'll notice addXrpLogo function, this function is used to add the XRP logo to the pages. This is an optional step just for aesthetic purpose, you can skip this step if you don't want to add the XRP logo to the pages.

> {{ include_code("_code-samples/build-a-wallet/js/src/helpers/render-xrpl-logo.js", language="js") }}

Reviewer: Since this logo isn't part of the lesson being taught, I'd recommend just having the sample code handle it automatically if possible (Or including it in the initial setup so it's not something they pay attention to)

Author: Removed this bit

> 7. Writing the HTML and CSS for the home page. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html) file to your index.html file.
> 
> It's basic HTML code with preloading CSS files, we are using plain CSS for this tutorial. You can use any CSS framework of your choice and beautify the page as you like. For now, we'll keep it simple. The CSS code is written in the index.css file. You can copy the code from the [index.css](_code-samples/build-a-wallet/js/index.css) file to your project's index.css file.

Reviewer: I'd recommend separating out the narrative/context building text from the instructions. That way people don't lose track of the action item.

```suggestion
Up next we need to update the HTML and CSS for the home page to give it the XRPL-look and feel. 

7. Copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html) file to your index.html file.

For this tutorial, we're just using basic HTML and pre-loaded CSS files. If you'd like to change the aesthetic, you can replace it with any CSS framework of your choice!

8. Copy the code from the [index.css](_code-samples/build-a-wallet/js/index.css) file to your project's index.css file to use the pre-made stylings.
```

### Step 3: Creating the Send XRP Page

> Till now, we have created the home page that displays the account and ledger details. In this step, we will create the Send XRP page that will allow users to send XRP to another account with optional destination tag.

Reviewer: The destination tag is an implementation detail and at this point is still jargon we don't expect the reader to understand, so I'd recommend removing it from the high level summary.
```suggestion
Till now, we have created the home page that displays the account and ledger details. In this step, we will create the Send XRP page that will allow users to send XRP to another account.
```

> ## Pre-requisites

Author: @JST5000 Can you guide me on how I can compile this md file to view it in the browser. Was having trouble when I followed the links you provided beforehand.

Reviewer: Hmmm... I also can't seem to figure out where the tutorial is generated. @mDuo13 do you know if there's a special command to generate tutorial files like `build-a-desktop-wallet-in-python.md`? (After doing `dactyl_build` I'm unable to find a file with the same name, or even the same text inside of it)

Reviewer: (As an aside, if you're using VSCode, I just found out about the Live Server extension and it seems pretty handy for viewing HTML files - https://stackoverflow.com/a/48721885 )

Reviewer 2: > Hmmm... I also can't seem to figure out where the tutorial is generated. @mDuo13 do you know if there's a special command to generate tutorial files like `build-a-desktop-wallet-in-python.md`? (After doing `dactyl_build` I'm unable to find a file with the same name, or even the same text inside of it)

What machines do you guys use? After installing some libraries (xml, etc.) it works flawlessly on a Mac M1.

Reviewer: @TusharPardhe ok, figured it out. 

You need to do two things:

1. Add the tutorial to `dactyl-config.yml` **(Screenshot of how attached)**
2. Add the header from the [comment below](https://github.com/XRPLF/xrpl-dev-portal/pull/1752#discussion_r1119446181) to your tutorial.md file (copied from the Python tutorial w/ the name changed from "Python" to "JavaScript")

Then when you run `dactyl_build` I see this error, which seems to mean it's actually compiling it now :)
```
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/dactyl/page.py", line 218, in preprocess
    md = self.pp_template.render(**context, **self.filter_exports())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "content/tutorials/build-apps/build-a-desktop-wallet-in-js.md", line 98, in top-level template code
  File "/Users/jacksonmills/Documents/bounty_submissions/tushar-dev-portal/tool/filter_include_code.py", line 9, in include_code
    with open("content/"+filename, encoding="utf-8") as f:
Preprocessor error in page <Page 'Pre-requisites' (build-a-desktop-wallet-in-js.html)>: [Errno 2] No such file or directory: 'content/_code-samples/build-a-wallet/js/src/helpers/render-xrp-logo.js'.
```

<img width="807" alt="image" src="https://user-images.githubusercontent.com/19694186/221956587-14aeb3b7-bf5e-40fc-b6bd-66802053081b.png">

Reviewer: @AlexanderBuzz I think what happened for me was I didn't delete the `out` folder, and it didn't fully update for whatever reason. After clearing out that file, I was able to generate the Python version, and after following the above steps it seemed to include the js version as well (although the build failed)

Author: Thanks @AlexanderBuzz I was able to build the webpage.

@JST5000 I forgot to push the changes that's why the build failed, please try again it will work now :)

2. Run 
   ```sh
    yarn create vite-app <project-name>
    ```
    or
    ```sh
    npm init vite-app <project-name>
    ```
    to create a new project.

3. Navigate to your project directory and copy this [package.json](_code-samples/build-a-wallet/js/package.json) file to your project.
        - Alternatively you can also do `yarn add <package-name>` or `npm install <package-name>` for each individual package to add them to your `package.json` file.
  
4. Run `yarn` or `npm i` to install the packages.

5. Create a new file `.env` in the root directory of the project and add the following variables:

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"
SEED="sEd7B2Jm7RBhhh5i88SH38aZB8ot8Kq" // Replace with your seed
```
6. Change the seed to your own seed. You can get a testnet account from [here](https://xrpl.org/xrp-test-net-faucet.html).

### Step 2: Creating the Home Page (Displaying Account & Ledger Details)

In this step, we will create the home page that will display the account details and the ledger details.

![Home Page Screenshot](img/js-wallet-home.png)

> - In your project directory, create new files named index.html, index.js and index.css. You can copy the contents of the [index.html](_code-samples/build-a-wallet/js/index.html), [index.js](_code-samples/build-a-wallet/js/index.js) and [index.css](_code-samples/build-a-wallet/js/index.css) files to your project's files.

Reviewer: Similar to above - from here on out I'd recommend using numbers instead of bullets whenever you want someone to take action. Otherwise it kinda looks like notes rather than action items. (And it can be hard to find where you left off if you don't have a number to come back to)

Author: noted

> - Make a new folder named src in the root directory of the project and add two folders named assets and helpers. In assets folder add contents of the [logo.png](_code-samples/build-a-wallet/js/src/assets/logo.png) to render the logo on the home page.

Reviewer: Split up any "and" into it's own step and add backtick quotes around filenames to make them look like literal names rather than normal words. (These rules should apply going forward)
```suggestion
- Make a new folder named `src` to the root directory of the project
- Add two folders named `assets` and `helpers` under `src`. 
- In the `assets` folder add contents of the [logo.png](_code-samples/build-a-wallet/js/src/assets/logo.png) to render the logo on the home page.
```

## Prerequisites

To implement this tutorial you should have a basic understanding of JavaScript and Node.js. You should also have a basic idea about XRP Ledger. For more information, visit the [XRP Ledger Dev Portal](https://xrpl.org) and the [XRPL Learning Portal](https://learn.xrpl.org/) for videos, libraries, and other resources.

Follow the steps below to get started:

1. [Node.js](https://nodejs.org/en/download/) (v10.15.3 or higher)
2. Install [Yarn](https://yarnpkg.com/en/docs/install) (v1.17.3 or higher) or [NPM](https://www.npmjs.com/get-npm) (v6.4.1 or higher)

## Source Code

You can find the complete source code for all of this tutorial's examples in the [code samples section of this website's repository]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/js/).

## Goals

At the end of this tutorial, you should be able to build a simple XRP wallet displayed below.

![Home Page Screenshot](img/js-wallet-home.png)

> This application is capable of the following:
> 
> - Shows updates to the XRP Ledger in real-time.
> - Can view any XRP Ledger account's activity "read-only" including showing how much XRP was delivered by each transaction.
> - Shows how much XRP is set aside for the account's reserve requirement. 
> - Can send direct XRP payments, and provides feedback about the intended destination address, including:
>   - Displays available balance in your account
>   - Verifies that the destination address is valid
>   - Validates amount input to ensure it is a valid number and that the account has enough XRP to send
>   - Allows addition of the destination tag

Reviewer: Simplifying language & using a consistent tense for the list. (This is different than the python one, but I think makes it easier to read)

Also added links to the relevant docs page with more information.

```suggestion
This application can:

- Show updates to the XRP Ledger in real-time.
- View any XRP Ledger account's activity, including showing how much XRP was delivered by each transaction.
- Show how much XRP is set aside for the account's [reserve requirement](reserves.html). 
- Send [direct XRP payments](direct-xrp-payments.html), and provide feedback about the intended destination address, including:
    - Displaying your account's available balance
    - Verifying that the destination address is valid
    - Validating the account has enough XRP to send
    - Allowing you to specify a destination tag
```

## Steps

Before you begin, make sure you have the pre-requisites installed. Check your node version by running `node -v`. If you don't have node installed, you can download it [here](https://nodejs.org/en/download/).

### Step 1: Setting up the project

- Navigate to the directory that you want to create the project in.
- Run `yarn create vite-app <project-name>` or `npm init vite-app <project-name>` to create a new project.

> - Install the dependencies that are mentioned below:

Reviewer: Instead of bullets, these should be numbered since they are expected to be followed in order. 

I'd also recommend breaking the `yarn create vite-app <project-name>` /npm version into their own lines with "```sh ```" blocks to make a box that's copyable. That way people can easily tell what the exact command should be.

```json
{
    "@esbuild-plugins/node-globals-polyfill": "^0.2.3",
    "crypto-browserify": "^3.12.0",
    "events": "^3.3.0",
    "https-browserify": "^1.0.0",
    "rollup-plugin-polyfill-node": "^0.12.0",
    "stream-browserify": "^3.0.0",
    "stream-http": "^3.2.0",
    "vite": "^4.1.0",
    "dotenv": "^16.0.3",
    "xrpl": "^2.7.0-beta.2",
    "xrpl-tagged-address-codec": "^0.2.1"
}
```

> you can install them by copying the dependencies from the [package.json](_code-samples/build-a-wallet/js/package.json) file to your project's package.json and running `yarn` or `npm i`. You can also do `yarn add <package-name>` or `npm install <package-name>` for each individual package.

Reviewer: (Added numbers, but please make the numbers line up with the rest of the steps :) )
```suggestion
1. Copy this [package.json](_code-samples/build-a-wallet/js/package.json) file to your project.
        - Alternatively you can also do `yarn add <package-name>` or `npm install <package-name>` for each individual package to add them to your `package.json` file.
2. Run `yarn` or `npm i` to install the packages.
```

> - Create a new file `.env` in the root directory of the project and add the following variables:

Reviewer: Should also be numbered

```bash
CLIENT="wss://s.altnet.rippletest.net/" // or any other server
EXPLORER_NETWORK="testnet" // or "mainnet"

> SEED="sEd7B2Jm7RBhhh5i88SH38aZB8ot8Kq" // Replace with your seed

Reviewer: If you want them to change the seed, I'd recommend adding another step to do that. (It's really easy to copy/paste without reading too deeply)

---
parent: build-apps.html
targets:
  - en
  - ja # TODO: translate this page
blurb: Build a graphical desktop wallet for the XRPL using Javascript.
---
# Build a Desktop Wallet in Javascript
<!-- STYLE_OVERRIDE: wallet -->

This tutorial demonstrates how to build a desktop wallet for the XRP Ledger using the Javascript programming language and various libraries. This application can be used as a starting point for building a more complete and powerful application, as a reference point for building comparable apps, or as a learning experience to better understand how to integrate XRP Ledger functionality into a larger project.

## Prerequisites

> To implement this tutorial you should have a basic understanding of JavaScript and Node.js. You should also have a basic idea about XRP Ledger. For more information, visit the [XRP Ledger Dev Portal](https://xrpl.org) and the [XRPL Learning Portal](https://learn.xrpl.org/) for videos, libraries, and other resources.
> 
> Follow the steps below to get started:
> 
> 1. [Node.js](https://nodejs.org/en/download/) (v10.15.3 or higher)
> 2. Install [Yarn](https://yarnpkg.com/en/docs/install) (v1.17.3 or higher) or [NPM](https://www.npmjs.com/get-npm) (v6.4.1 or higher)

Reviewer: Changing the general links to be a specific recommendation of a tutorial to do before starting this one.

```suggestion
To complete this tutorial, you should meet the following guidelines:

1. You have [Node.js](https://nodejs.org/en/download/) v10.15.3 or higher installed.
2. You have [Yarn](https://yarnpkg.com/en/docs/install) (v1.17.3 or higher) or [NPM](https://www.npmjs.com/get-npm) (v6.4.1 or higher) installed.
3. You are somewhat familiar with coding with JavaScript and have completed the [Get Started Using JavaScript](get-started-using-python.html) tutorial.
```

> 3. Add your Seed, Client, and specify testnet/mainnet in .env file. Example .env file is provided in the source code below.
> 4. Run `yarn install` or `npm install` to install dependencies
> 5. Start the app with `yarn dev` or `npm dev`

Reviewer: These aren't pre-requisites because you'll walk them through doing this as part of the tutorial. We can just get rid of this for now.
```suggestion
```
